### DISCUSSION TITLE
extent() replacement in new master

### DISCUSSION THREAD

**Discussion by psiha** [jfalcou/kiwaku#85] [Q&A]
Up until now I had something like:
```
template < auto ... dims >
struct Dimensions : decltype( kwk::of_size( Detail::kwkDim< dims >()... ) )
{
   using Base = decltype( kwk::of_size( Detail::kwkDim< dims >()... ) );

   template < unsigned dimIndex >
   constexpr auto getValue() const noexcept { return this-> template extent< dimIndex >(); }
}
```

this no longer compiles, I see that prefilled does/tries friend injection for a free standing get<> (into the kwk::__ namespace?) however I can't get that to compile, neither of these compile (no mathing member function for call to get):
```
{ return get< dimIndex >( static_cast< Base const & >( *this ) ); }
{ return std::get< dimIndex >( static_cast< Base const & >( *this ) ); }
```

**Comment by psiha** [comment]
@jfalcou don't know if you get automatically notified for new 'discussions' 😊

**Reply by jfalcou** [reply]
I am. Will answer tomorrow.

**Comment by psiha** [comment]
king ping 😁

**Reply by jfalcou** [reply]
Quick fix, try kumi::get explictly

Now, I think that our changes on this front should make your adapter more or less superfluous. LEt's catch up in chat and discuss this deeper

**Reply by psiha** [reply]
doesn't help :/
```
 error : no matching function for call to 'get'
 note: candidate template ignored: could not match 'tuple' against 'prefilled'
```

**Reply by psiha** [reply]
what would be your intedend way of use here?

even if the adapter is hopefully no longer necessary such usage should still work?

**Reply by psiha** [reply]
kingping#2 :P i'm stuck with this/show stopper - i need a constexpr getter of values from shapes 'n' strides

**Reply by jfalcou** [reply]
OK, this works:

https://godbolt.org/z/vsKhsTzfq

Tell me if it does for you.
If it doesn't work for you, then something is maybe wrong with your Detail::kwkDim.

**Reply by jfalcou** [reply]
If you want 0 to gives you the dimension 0 and not the 0th element of the tuple then:

https://godbolt.org/z/8Tvszfnev

**Reply by psiha** [reply]
unfortunately it doesn't work for me - still the same error.

kwkDim is:
```
    template < auto mbDim >
    constexpr auto kwkDim() noexcept
    {
        auto constexpr size{ extent( mbDim ) };
        if constexpr ( size == Dimension::dynamicValue )
            return size;
        else
            return kwk::fixed< size >;
    }
```
and it could be the culprit - as i see there have been substantial changes here on your side - Dimension::dynamicValue is 0 - and so for a dynamic axis [where you would use extent() - i.e. with the paren-paren call operator - to define it) it returns 0 - this used to work before as you also used that value for dynamic axes - what do i need to return now?

**Reply by jfalcou** [reply]
o_O I am confused on how it could have worked before. extent() never had a parameter.
Can you show me an example of cal to Dimensions with different values so I cna see wat is the correct way to do it rn

**Reply by psiha** [reply]
ok i get it, figures 🙈 the problem is i have a member named get (used for fetching a dimension by type rather by index) - this gets you into this situation: https://godbolt.org/z/6Y9xW6P36
what namespace is your free get supposed to be at?

ps. you have compilation issues with clang-trunk

**Reply by jfalcou** [reply]
There supposed to be in kwk or kwk::detail.
Using

`return (*this)[kwk::fixed<dimIndex>];`

fix the compilation it seems.

PS: yeah, maybe, it's trunk so until it is not 16, I will wait for them to fix stuff.

**Comment by psiha** [comment]
> There supposed to be in kwk or kwk::detail.

none of kwk::, kwk::__, kwk::detail work :/ 🤷
will file a separate issue

> return (*this)[kwk::fixed<dimIndex>];

ok thanks this fixed most of the issues - still one weird remains (investigating)...

> PS: yeah, maybe, it's trunk so until it is not 16, I will wait for them to fix stuff.

maybe there will be fixes on your side too, like a missing cstdint include directive ;D

**Reply by jfalcou** [reply]
Will look at the issue tonight.

As for the missing cstdint, fine you got me :p

**Reply by psiha** [reply]
ps. do you have an idea of when you'll be open for gmeet/zoom/discord session?
+ping about the cppp email ;)
